llvm/test/Transforms/LoopVectorize/PowerPC/vplan-force-tail-with-evl.ll

```diff
@@ -21,7 +21,31 @@ define void @foo(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %N) {
 ; CHECK-NEXT:   vector.body:
 ; CHECK-NEXT:     EMIT vp<[[CAN_IV:%.+]]> = CANONICAL-INDUCTION ir<0>, vp<[[CAN_INC:%.*]]>
 ; CHECK-NEXT:     WIDEN-INDUCTION %iv = phi 0, %iv.next, ir<1>, vp<[[VF]]>
+; CHECK-NEXT:     vp<[[STEPS:%.+]]> = SCALAR-STEPS vp<[[CAN_IV]]>, ir<1>
 ; CHECK-NEXT:     EMIT vp<[[CMP:%.+]]> = icmp ule ir<%iv>, vp<[[BTC]]>
+; CHECK-NEXT:   Successor(s): pred.load
```

Comment:
At first glance this looks worse than before, unless I'm missing something. It looks like previously we were reusing the same predicated blocks to perform both the load and store, i.e. something like

```
  %cmp = icmp ... <4 x i32>
  %lane0 = extractelement <4 x i32> %cmp, i32 0
  br i1 %lane0, label %block1.if, label %block1.continue

%block1.if:
  .. do load ..
  .. do store ..
...
```

whereas now we've essentially split out the loads and stores with duplicate control flow, i.e.

```
  %cmp = icmp ... <4 x i32>
  %lane0 = extractelement <4 x i32> %cmp, i32 0
  br i1 %lane0, label %block1.load.if, label %block1.load.continue

%block1.if:
  .. do load ..
...

%stores:
  %lane0.1 = extractelement <4 x i32> %cmp, i32 0
  br i1 %lane0.1, label %block1.store.if, label %block1.store.continue
...
```

I'd expect the extra control flow to hurt performance.

Reply:
This test is a bit strange as both before and after this patch this VPlan isn't used, as it decides the scalar version costs less. As to why it decides this VPlan is the best for this vector length, computePredInstDiscount doesn't currently account for any control flow instructions it could be able to remove by scalarizing everything, though I don't know how it would do that, as it's making a decision about whether to scalarize the chain of instructions feeding into a store, and the chose to scalarize the loads happens elsewhere.